Feature: Support for Cloudtrail Management events

shyammukund opened the pull request:
### Summary

- Implemented CloudTrail management events module for AWS role assumption visibility


Key Features:
- Fetches STS events (AssumeRole, AssumeRoleWithSAML, AssumeRoleWithWebIdentity)
- Creates ASSUMED_ROLE relationships between principals and roles
- Handles role-to-role, user-to-role, and service-to-role assumptions
- Supports configurable lookback periods (default 24 hours)




### Related issues or links
> Include links to relevant issues or other pages.

- https://github.com/cartography-cncf/cartography/issues/...


### Checklist

Provide proof that this works (this makes reviews move faster). Please perform one or more of the following:
- [ ] -Unit Tests
<img width="1550" alt="image" src="https://github.com/user-attachments/assets/20396056-86b3-46ce-bb94-4fef7a59156c" />
- [ ] -Integration Tests
<img width="1553" alt="image" src="https://github.com/user-attachments/assets/c13bf5a7-8bf9-4a99-924a-2c9317a7da39" />

- [ ] Include a screenshot showing what the graph looked like before and after your changes.
<img width="1816" alt="image" src="https://github.com/user-attachments/assets/7561911e-0b37-4af0-a290-93ae073e9c75" />

- [ ] Cartography Log (**AWS account id replaced with placeholder**)
WARNING:cartography.cli:A Kandji base URI was not provided.
WARNING:cartography.cli:A SnipeIT base URI was not provided.
INFO:cartography.sync:Starting sync with update tag '1751323749'
INFO:cartography.sync:Starting sync stage 'aws'
INFO:cartography.intel.aws:Syncing AWS accounts: 0123456789012
INFO:cartography.intel.aws:Syncing AWS account with ID '0123456789012' using configured profile 'ReadOnlyAccess-0123456789012'.
INFO:cartography.intel.aws:Trying to autodiscover accounts.
WARNING:cartography.intel.aws:The current account (0123456789012) doesn't have enough permissions to perform autodiscovery.
INFO:cartography.intel.aws.cloudtrail_management_events:Starting CloudTrail management events sync for account 0123456789012
INFO:cartography.intel.aws.cloudtrail_management_events:Syncing 1 regions with 24 hour lookback period
INFO:cartography.intel.aws.cloudtrail_management_events:Fetching CloudTrail management events for region 'us-east-1' from 2025-06-29 22:49:12.635056 to 2025-06-30 22:49:12.635056 (24 hours)
INFO:cartography.intel.aws.cloudtrail_management_events:Retrieved 2400 CloudTrail management events from region 'us-east-1'
INFO:cartography.intel.aws.cloudtrail_management_events:Transforming 2400 CloudTrail events to role assumptions for region 'us-east-1'
WARNING:cartography.intel.aws.cloudtrail_management_events:Failed to transform CloudTrail event 83d124e3-d114-43ac-a0ea-f89432bb365a: 'NoneType' object has no attribute 'get'
WARNING:cartography.intel.aws.cloudtrail_management_events:Failed to transform CloudTrail event 8f5e5304-34f2-456b-b4a4-33c56ec1881b: 'NoneType' object has no attribute 'get'
INFO:cartography.intel.aws.cloudtrail_management_events:Successfully transformed 2356 role assumptions from 2400 events
INFO:cartography.intel.aws.cloudtrail_management_events:Aggregated 2356 events into 12 unique role assumption relationships
INFO:cartography.intel.aws.cloudtrail_management_events:Loading 12 aggregated role assumption relationships
INFO:cartography.intel.aws.cloudtrail_management_events:Successfully loaded 12 role assumption relationships
INFO:cartography.intel.aws.cloudtrail_management_events:CloudTrail management events sync completed successfully. Processed 2356 role assumption events.
INFO:cartography.graph.statement:Completed aws_ec2_iaminstanceprofile statement #1
INFO:cartography.graph.statement:Completed aws_ec2_iaminstanceprofile statement #2
INFO:cartography.graph.job:Finished job aws_ec2_iaminstanceprofile
INFO:cartography.graph.statement:Completed aws_lambda_ecr statement #1
INFO:cartography.graph.statement:Completed aws_lambda_ecr statement #2
INFO:cartography.graph.job:Finished job aws_lambda_ecr
INFO:cartography.graph.statement:Completed aws_post_ingestion_principals_cleanup statement #1
INFO:cartography.graph.job:Finished job aws_post_ingestion_principals_cleanup
INFO:cartography.util:Did not run aws_ec2_asset_exposure.json because it needs {'ec2:security_group', 'ec2:load_balancer', 'ec2:load_balancer_v2', 'ec2:instance'} to be included as a requested sync. You specified: {'cloudtrail_management_events'}. If you want this job to run, please change your CLI args/cartography config so that all required resources are included.
INFO:cartography.util:Did not run aws_ec2_keypair_analysis.json because it needs {'ec2:keypair'} to be included as a requested sync. You specified: {'cloudtrail_management_events'}. If you want this job to run, please change your CLI args/cartography config so that all required resources are included.
INFO:cartography.util:Did not run aws_eks_asset_exposure.json because it needs {'eks'} to be included as a requested sync. You specified: {'cloudtrail_management_events'}. If you want this job to run, please change your CLI args/cartography config so that all required resources are included.
INFO:cartography.graph.statement:Completed aws_foreign_accounts statement #1
INFO:cartography.graph.statement:Completed aws_foreign_accounts statement #2
INFO:cartography.graph.job:Finished job aws_foreign_accounts
INFO:cartography.sync:Finishing sync stage 'aws'
INFO:cartography.sync:Finishing sync with update tag '1751323749'

If you are changing a node or relationship:
- [ ] Update the [schema](https://github.com/cartography-cncf/cartography/tree/master/docs/root/modules) and [readme](https://github.com/cartography-cncf/cartography/blob/master/docs/schema/README.md).

If you are implementing a new intel module:
- [ ] Use the NodeSchema [data model](https://cartography-cncf.github.io/cartography/dev/writing-intel-modules.html#defining-a-node).

achantavy commented on cartography/intel/aws/__init__.py:
```diff
@@ -53,7 +54,7 @@ def _sync_one_account(
     update_tag: int,
     common_job_parameters: Dict[str, Any],
-    regions: list[str] | None = None,
+    regions: Union[List[str], None] = None,
```
Use the python 3.10 typing style of

`list[str] | None` instead of `Union[List[str], None]`. 

Older code uses the older style, but newer code should use the newer one whereever possible

achantavy commented on cartography/intel/aws/__init__.py:
```diff
@@ -177,7 +178,7 @@ def _sync_multiple_accounts(
     aws_best_effort_mode: bool,
     aws_requested_syncs: List[str] = [],
-    regions: list[str] | None = None,
+    regions: Union[List[str], None] = None,
```
param style

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+
+            all_role_assumptions.extend(role_assumptions)
+
+        except Exception as e:
+            logger.warning(
+                f"Failed to process CloudTrail events for region {region}: {str(e)}"
+            )
+            continue
```
Remove the try + except here. This isn't adding much value

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+                current_aws_account_id=current_aws_account_id,
+            )
+
+            all_role_assumptions.extend(role_assumptions)
```
Put the load in here so that we load in the events region by region. There will likely be TONS of events so aggregating them all across all regions may exhaust memory. (It might even be too much to carry for a single region but we'll figure that out later)

shyammukund replied:
put load in the loop so it loads region by region

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+        load_role_assumptions(
+            neo4j_session=neo4j_session,
+            role_assumptions=all_role_assumptions,
+            region="all",  # Indicates cross-region aggregation
```
What is `'all'`? There is no such thing as an `'all'` region. Remove this param.

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+def load_role_assumptions(
+    neo4j_session: neo4j.Session,
+    role_assumptions: List[Dict[str, Any]],
+    region: str,
```
This parameter is not used -- remove

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+        )
+
+        # Run cleanup for stale relationships
+        matchlink_schema = AssumedRoleMatchLink()
+        cleanup_job = GraphJob.from_matchlink(
+            matchlink_schema,
+            "AWSAccount",
+            current_aws_account_id,
+            update_tag,
+        )
+        cleanup_job.run(neo4j_session)
```
Move this to a separate cleanup function like the other modules.

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+            f"CloudTrail management events sync completed successfully. "
+            f"Processed {len(all_role_assumptions)} role assumption events."
+        )
+    else:
+        logger.info(
+            "CloudTrail management events sync completed - no role assumptions found"
+        )
```
Remove this if+else. It's not adding much and just makes the code look more nested and complicated

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+            f"Retrieved {len(events)} CloudTrail management events from region '{region}'"
+        )
+
+    except Exception as e:
+        logger.warning(
+            f"Failed to retrieve CloudTrail management events for region '{region}': {str(e)}"
+        )
```
Don't handle exceptions. `@aws_handle_regions` does that for you already

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+    )
+
+    for event in events:
+        try:
```
Have you run the code and ran into this exception? If not, remove the try+except. Pre-emptive exceptions are not needed for this scenario. Code for the happy path and add exceptions as needed.

Right now, even if this exception does happen, we're writing a log message and swallowing it, and this makes the code harder to trace if that issue comes up. Let the exception fail loudly so that it's easier to trace. Then we can add in the handling if appropriate in quick follow-up PRs.

achantavy replied:
[block] can we remove the try + excepts here?

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+    :param aws_update_tag: Timestamp tag for tracking data freshness
+    :rtype: None
+    """
+    if not role_assumptions:
+        logger.info("No role assumption events to load")
+        return
```
Not needed because `load()` already handles the empty list case.

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+        return
+
+    # Aggregate role assumptions by (source, destination) pairs
+    aggregated_assumptions = _aggregate_role_assumptions(role_assumptions)
```
This should be in the transform step. the `load()` step should not do transforms

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+        _sub_resource_id=current_aws_account_id,
+    )
+
+    logger.info(
+        f"Successfully loaded {len(aggregated_assumptions)} role assumption relationships"
+    )
```
Not needed

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+        event_time = assumption.get("EventTime")
+
+        if not source_arn or not dest_arn or not event_time:
+            logger.warning(
+                "Skipping incomplete assumption record: due to missing required fields"
+            )
```
Think about the error here. If we have incomplete data, what do we want to do with it? What other information do we need to know about in production? Should we include other fields? Or should we not include this data at all? Right now if I got this message in a log system I'd just get a bunch of noise and I would have no idea what to do about it, and I wouldn't know what data I'd potentially be missing.

shyammukund replied:
I think it would be helpful to have this info but agree that just saying a field is missing isnt helpful at all. I changed so that we check which fields are missing and state those fields in the logger warning.

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+        agg_data["lastused"] = agg_data["last_seen"]
+        result.append(agg_data)
+
+    logger.info(
+        f"Aggregated {len(role_assumptions)} events into {len(result)} unique role assumption relationships"
```
Not needed

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+        # Focus on STS events for role assumptions
+        paginator = client.get_paginator("lookup_events")
+
+        page_iterator = paginator.paginate(
```
Is there a way to filter the event name from the API in the `get` function so that we don't need to do it in the transform function later? This would let us save on memory

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+    event_name = event.get("EventName")
+
+    # Only process STS role assumption events
+    if event_name not in [
```
See above comment about checking on if its possible to filter for these events at the API level in the get instead of in the transform

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+    return result
+
+
+def _extract_role_assumption_from_event(
```
I think these `_extract*()` functions are well-intentioned but is more complicated than what it needs to be. We can likely get away with just direct dictionary accesses like with `mydata['keyname']['keyname2']`.

shyammukund replied:
I went ahead and changed the source and destination principals to use direct dict access to extract instead of their own helper funcs. Now transform contains all this logic (source and dest principal extraction + json parsing)

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+    role_assumptions: List[Dict[str, Any]],
+) -> List[Dict[str, Any]]:
+    """
+    Aggregate role assumption events by (source_principal, destination_principal) pairs.
```
We should think a bit more on an optimal approach here since this is likely to be a memory bottleneck

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+    return assumption
+
+
+def _extract_source_principal(event: Dict[str, Any]) -> Optional[str]:
```
Same point about this looking overly complicated. Can we not get away with just direct dict accesses?

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+    return None
+
+
+def _extract_destination_principal(event: Dict[str, Any]) -> Optional[str]:
```
Direct dict accesses?

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,553 @@
+    }
+
+    # Add additional context from CloudTrail event JSON
+    cloudtrail_event = _parse_cloudtrail_event_json(event.get("CloudTrailEvent"))
```
Double checking, the data from this field is a string represented as json? It is not a python object already?

shyammukund replied:
yes cloudtrail_event contains a string as json, but then im converting it to a python dict using json.load(). I refactored this to just be in the transform function so we only have to parse once per event

achantavy reviewed (changes requested):
quick first pass on error handling and sharing some initial thoughts on log processing

shyammukund commented:
@achantavy addressed your comments on the last commit. Removed a lot of the logging and try +except blocks and left comments on the stuff I refectored/changed

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,472 @@
+                "source_principal_arn": normalized_source_arn,
+                "destination_principal_arn": dest_arn,
+                "times_used": 1,
+                "first_seen": event_time,
```
This should be named something like "first seen in time window' because it could be confused with the `firstseen` field. We don't know when this field was first seen, we only know when in the given time window it was seen.

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,472 @@
+        if event.get("CloudTrailEvent"):
+            try:
+                cloudtrail_event = json.loads(event["CloudTrailEvent"])
+            except (json.JSONDecodeError, TypeError):
+                logger.debug(
+                    f"Failed to parse CloudTrail JSON for event {event.get('EventId', 'unknown')}"
+                )
```
For a first PR on this, let's not apply this level of defensive programming yet. Let the exception fail.

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,472 @@
+
+        # Simple source principal extraction
+        user_identity = event.get("UserIdentity", {})
+        source_principal = (
+            user_identity.get("arn")
+            or user_identity.get("principalId")
+            or event.get("UserName")
+            or (
+                cloudtrail_event and cloudtrail_event.get("userIdentity", {}).get("arn")
+            )
+        )
+
+        if not source_principal:
+            logger.debug(
+                f"Could not extract source principal from event {event.get('EventId', 'unknown')}"
+            )
+            continue
```
is it really this hard and complex to get a source principal out? Pick one place to get it from and that's it. there is no need to use this many fallbacks at such an early stage of the code.

From the [spec](https://docs.google.com/document/d/1GvOeLB6zC1l0-PPn7wG-etybrP0bwlheKarxkJ_5_bs/edit?tab=t.0#heading=h.bg8k06ncfpe5), I think that's in `requestParameters['roleArn']`.

If it's none or whatever, so be it, let it fail, we'll fix it in a follow up

shyammukund replied:
I was able to simplify source_principal to just use simple dict access. 

Im having a little more trouble with getting dest principal to have similar super simple access. Right now what im doing is iterating over the events resources and if i find a ResourceType that equals AWS:IAM:Role then I use dict access to assign that to destination_principal, but this still seems like it is too complex. 
  
Also when i run cartography an error pops up that says that some events are missing fields so I put a try except in this section so that it skips those events and continues the sync.

It would be helpful if we could look at this section of code together to see where I'm going wrong.

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,472 @@
+
+        # Simple destination principal extraction
+        destination_principal = None
+        if cloudtrail_event:
+            # Try request parameters first
+            destination_principal = cloudtrail_event.get("requestParameters", {}).get(
+                "roleArn"
+            )
+
+            # Fallback: try response elements
+            if not destination_principal:
+                assumed_role_arn = (
+                    cloudtrail_event.get("responseElements", {})
+                    .get("assumedRoleUser", {})
+                    .get("arn")
+                )
+                if assumed_role_arn:
+                    destination_principal = _convert_assumed_role_arn_to_role_arn(
+                        assumed_role_arn
+                    )
+
+        if not destination_principal:
+            logger.debug(
+                f"Could not extract destination principal from event {event.get('EventId', 'unknown')}"
+            )
+            continue
```
Again, let's pick one place and not use fallbacks. From the spec, we can at minimum get a principal ID from `theobject['userIdentity']['principalId']`. We need to do a bit of parsing to get that id, and then we have to match that role id on an `:AWSRole`'s `roleid` field.

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,472 @@
+    )
+
+
+def _aggregate_role_assumptions(
```
As mentioned above, this really should be the main logic of the `transform()` function since these are the only fields that are really needed to load things to the graph.

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,472 @@
+
+            # Construct the IAM role ARN
+            return f"arn:aws:iam::{account_id}:role/{role_name}"
+    except (IndexError, AttributeError):
+        logger.debug(
+            f"Could not convert assumed role ARN to role ARN: {assumed_role_arn}"
+        )
```
Don't catch unless you've actually seen it happen before

achantavy reviewed (changes requested):
Comments on consolidating and efficiency and not being overly defensive in the code. Rule of thumb (for now): don't handle an exception unless you've actually seen it happen!

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,339 @@
+        normalized_destination_principal = _convert_assumed_role_arn_to_role_arn(
+            destination_principal
+        )
+        event_time = event.get("EventTime")
```
[block] event time can be None. If this happens then the `<` and `>` checks below will crash.

i would do  somethign like

```python
if not event_time:
    aggregated[key]["first_seen_in_time_window"] = None
    aggregated[key]["last_used"] = None
else:
    aggregated[key]["first_seen_in_time_window"] = min(aggregated[key]["first_seen_in_time_window"], event_time)
    aggregated[key]["last_used"] = max(aggregated[key]["last_used"], event_time)
```

Note also that the use of min and max above helps consolidate a little bit, but that part of suggestion is optional

achantavy commented on cartography/intel/aws/cloudtrail_management_events.py:
```diff
@@ -0,0 +1,339 @@
+                "times_used": 1,
+                "first_seen_in_time_window": event_time,
+                "last_used": event_time,
+                "event_types": ["AssumeRole"],
+                "assume_role_count": 1,
+                "saml_count": 0,
+                "web_identity_count": 0,
```
[block]  Remove these fields, they're not used  because this sync function is only for assumerole

achantavy reviewed (changes requested):
almost there!!
